Towards #385. PSyIR Fortran front-end Do_Constructs handler.

sergisiso opened the pull request:
This PR:
- Adds a generic Fortran do loop handler to the ASTProcessor based on the implementation of the NemoLoop parser.
- start/stop/step expressions of Loop are now in-tree. (Just the do_loop bounds)
```
Loop[type='None', field_space='None', it_space='None']
        Literal[value:'1']   # Child 1 is the Start bound
        Reference[name:'un'] # Child 2 is the Stop bound
        Literal[value:'1'] # Child 3 is the Stepping value
        Schedule[]  # Child 4 is the Loop body (always a Schedule)
            Assignment[]
                Reference[name:'rtmp1']
                Literal[value:'3']
```
This is still not able to represent different types of Loops, it is limited to Fortran Do Loops.

sergisiso commented:
@rupertford It won't pass the tests & coverage checks yet but the fparser -> PSyIR should already work, could you try if this already covers the LFRic kernels Loops?

Also, at the moment it generates generic `Loop` nodes and I will make it generate `NemoLoops` for the nemo API, but should LFric kernels generate `DynLoops`? or do `DynLoops` semantics only make sense in the PSy layer?

rupertford commented:
Hi @sergisiso, I've run the Kernel-to-PSyIR translation on one of the LFRic algorithm codes (the first alphabetically).

```
psyclone -s ./kernel_print.py ../../../../lfric_trunk/gungho/source/algorithm/advection_alg_mod.x90 -d ../../../../lfric_trunk/gungho/source/kernel -oalg /dev/null -opsy /dev/null > res.out
```

where kernel_print.py is

```
from __future__ import print_function
def trans(psy):
    '''Print out Fortran versions of all kernels found in this file.'''

    # Loop over all of the Invokes in the PSy object.
    for invoke in psy.invokes.invoke_list:
        schedule = invoke.schedule
        for kern in schedule.kern_calls():
            kern_schedule = kern.get_kernel_schedule()
            kern_schedule.view()

    return psy
```

It's looking good but there are a couple of errors

1) one of the view() methods needs an int to be converted to a string.
```
diff --git a/src/psyclone/psyGen.py b/src/psyclone/psyGen.py
index 3b1ff50d..126947c4 100644
--- a/src/psyclone/psyGen.py
+++ b/src/psyclone/psyGen.py
@@ -7198,7 +7198,7 @@ class Literal(Node):
         :param int indent: level to which to indent output.
         '''
         print(self.indent(indent) + self.coloured_text + "["
-              + "value:'"+self._value + "']")
+              + "value:'"+str(self._value) + "']")
 
     def __str__(self):
         return "Literal[value:'" + self._value + "']\n"
```
2) Not all of the kernel tree is being parsed for some reason.
Here is an example:

```
  do k = 0, nlayers-1
    do df = 1, ndf2  
      x_e(df) = x(map2(df)+k)
    end do
    ik = (cell-1)*nlayers + k + 1
    lhs_e = matmul(matrix(:,:,ik),x_e)
    do df = 1,ndf1
       lhs(map1(df)+k) = lhs_e(df) 
    end do
  end do
```

gives

```
Schedule[name:'dg_matrix_vector_code']
    Loop[type='None',field_space='None',it_space='None']
        Literal[value:'0']
        BinaryOperation[operator:'-']
            Reference[name:'nlayers']
            Literal[value:'1']
        Literal[value:'1']
        Schedule[]
            Loop[type='None',field_space='None',it_space='None']
                Literal[value:'1']
                Reference[name:'ndf2']
                Literal[value:'1']
                Schedule[]
                    Assignment[]
                        ArrayReference[name:'x_e']
                            Reference[name:'df']
                        ArrayReference[name:'x']
                            BinaryOperation[operator:'+']
                                ArrayReference[name:'map2']
                                    Reference[name:'df']
                                Reference[name:'k']
```

which means the `ik` assignment, `lhs_e` assignment and last (`ndf1`) loop are missing.

sergisiso commented:
@rupertford I believe these two issues are fixed now

sergisiso commented:
@arporter @rupertford This is ready for review now. This PR introduces the new Loop structure where the start/stop/step are part of the PSyIR. It still has some limitations listed at #440

The PR turned out quite long, but a big part of it is the updates of `Loop.children` -> `Loop.loop_body` (to avoid writing `Loop.children[3].children` of the new structure) and the update of tests that used `__str__` and `view()` methods that had some Loop in it.

The codecov is not 100% (and thus the ci fails) because I touched gocean0p1 and dynamo0p1 code which had no tests, but I think all new code is covered.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3220,17 +3313,24 @@ def gen_code(self, parent):
             from psyclone.f2pygen import DoGen, DeclGen
-            do = DoGen(parent, self._variable_name, self._start, self._stop)
+            # Start_expr and stop_expr are generated with the FortranWriter
+            # backend, the rest of the loop with fpygen.
```
s/fpygen/f2pygen/

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3220,17 +3313,24 @@ def gen_code(self, parent):
+            fwriter = FortranWriter()
+            do = DoGen(parent, self._variable_name,
+                       fwriter(self.start_expr),
+                       fwriter(self.stop_expr))
```
What about the step or is that only handled by sub-classes?

sergisiso replied:
It wasn't on the original code :), but yes I added it

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6777,6 +6883,107 @@ def _ignore_handler(self, *_):
+    def _create_loop(self, parent, variable_name):
+        '''
+        Create a Loop instance. This is done outside _do_construct_handler
+        because some APIs may want to instantiate a specialized Loop.
```
s/specialized/specialised/ ;-)

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6777,6 +6883,107 @@ def _ignore_handler(self, *_):
+        Create a Loop instance. This is done outside _do_construct_handler
+        because some APIs may want to instantiate a specialized Loop.
+
+        :param parent: The parent of the node.
```
Lower-case start to `:param:` and `:return:` descriptions please.

sergisiso replied:
fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6777,6 +6883,107 @@ def _ignore_handler(self, *_):
+
+    def _process_loopbody(self, loop_body, node):
+        ''' Process the loop body. This is done outside _do_construct_handler
+        becayse some APIs may want to perform specialized actions. By default
```
s/becayse/because/ and s/specialized/specialised/

sergisiso replied:
fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6777,6 +6883,107 @@ def _ignore_handler(self, *_):
+        :param node: node in fparser2 tree.
+        :type node: \
+            :py:class:`fparser.two.Fortran2003.Block_Nonlabel_Do_Construct`
+        :param parent: Parent node of the PSyIR node we are constructing.
```
s/Parent/parent/

sergisiso replied:
fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -6777,6 +6883,107 @@ def _ignore_handler(self, *_):
+            default_step = Literal("1", parent=loop)
+            loop.addchild(default_step)
+
+        # Process loop body (ignore 'do' and 'end do' statements with [1:-1])
```
There's no `[1:-1]` in the code here.

sergisiso replied:
Ups moved the comment line to where the [1:-1] is

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -419,6 +419,35 @@ def ifblock_node(self, node):
+        '''This method is called when a Loop instance is found in the
+        PSyIR tree.
+
+        :param node: An Loop PSyIR node.
```
s/An Loop/a Loop/

sergisiso replied:
fixed

arporter commented on src/psyclone/psyir/backend/fortran.py:
```diff
@@ -419,6 +419,35 @@ def ifblock_node(self, node):
+            "{0}do {1} = {2}, {3}, {4}\n"
+            "{5}"
+            "{0}enddo\n"
+            "".format(self._nindent, variable_name, start, stop, step, body))
```
Nice. Presumably the body is indented correctly? (Something to do with `self._depth`?)

sergisiso replied:
yep, self._depth keeps track of the loop, for this reason it is increased and decreased around the `node.loop_body` visits.

arporter replied:
Thanks.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -119,6 +119,13 @@ def _validate(self, node_list):
                 if not isinstance(item, self.valid_node_types):
+                    # TODO: To the reviewer, Literals and References appear
+                    # everywhere now, should they be in all transformations
+                    # 'valid_node_types' or is it OK to skip them here. Also
```
Yes, I think they should all appear in `valid_node_types`. As @rupertford pointed out though, we (I) probably need to re-think this functionality. Possibly we could reverse it and
have an `excluded_node_types` list but that's not for this issue.

sergisiso replied:
OK, I added them to all `valid_node_type`s, but I support the `excluded_nodes` as a better solution (and more readable to understand the current restrictions of each transformation).

arporter commented on src/psyclone/transformations.py:
```diff
@@ -207,14 +214,16 @@ def _validate(self, node1, node2):
+        :raises TransformationError: if the nodes do not have the same parent.
+        :raises TransformationError: if the nodes are not next to each \
+            other in the tree.
+        :raises TransformationError: if the \
```
Thanks for fixing the formatting.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -250,11 +273,11 @@ def apply(self, node1, node2):
 
         # add loop contents of node2 to node1
-        node1.children.extend(node2.children)
+        node1.loop_body._children.extend(node2.loop_body)
```
pylint's not going to like the use of `_children`. Is there a `children` property we can use instead?

sergisiso replied:
Sure

arporter commented on src/psyclone/transformations.py:
```diff
@@ -250,11 +273,11 @@ def apply(self, node1, node2):
-        for child in node2.children:
-            child.parent = node1
+        for child in node2.loop_body:
+            child.parent = node1.children[3]
```
Could `children[3]` be `loop_body`? i.e. does `loop_body` return the `Schedule` of the loop body?

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1197,14 +1223,14 @@ def apply(self, node):
 
         # add contents of node to colour loop
-        colour_loop.children.extend(node.children)
+        colour_loop.loop_body._children.extend(node.loop_body)
```
Again, can we replace `_children` with `children` here?

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1197,14 +1223,14 @@ def apply(self, node):
-        for child in node.children:
-            child.parent = colour_loop
+        for child in node.loop_body:
+            child.parent = colour_loop.children[3]
```
`loop_body` instead of `children[3]`?

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1918,15 +1944,16 @@ def _validate(self, node, depth):
                 "In the Dynamo0p3RedundantComputation transformation apply "
                 "method the first argument is not a Loop")
-        # check loop's parent is the schedule, or its parent is a
+        # Check loop's parent is the InvokeSchedule, or its nested in a
```
"... or that it is nested in a ..."

sergisiso replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2182,17 +2207,17 @@ def apply(self, outer):  # pylint: disable=arguments-differ
-            child.parent = outer
+        # Move inner's schedule to outer:
+        outer.children[3] = inner.loop_body
+        inner.children[3] = Schedule()
```
Do you have to create this empty Schedule for the Loop internal-consistency checks? Please add a comment.

sergisiso replied:
It is simply to remove all current entries (an alternative to loop through the body and `del` each children). I have reorganized and commented the code.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2182,17 +2207,17 @@ def apply(self, outer):  # pylint: disable=arguments-differ
-            outer.children.append(child)
-            child.parent = outer
+        # Move inner's schedule to outer:
+        outer.children[3] = inner.loop_body
```
`loop_body` instead of `children[3]`?

sergisiso replied:
Not in this case as the loop_body setter is not implemented, and I think it shouldn't be as it should be always fixed as an Schedule. Does that make sense?

arporter replied:
Yes it does. Thanks for explaining.

sergisiso commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -109,8 +109,11 @@ def create(call, parent=None):
+        # As it is the innermost loop it has the kernel as a loop_body
+        # child.
+        dofloop.loop_body.addchild(builtin)
+        builtin.parent = dofloop.children[3]
```
Yep

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -425,7 +425,7 @@ def __init__(self, parent=None,
 
         Loop.__init__(self, parent=parent,
-                      valid_loop_types=VALID_LOOP_TYPES)
+                      valid_loop_types=VALID_LOOP_TYPES,)
```
Removed

sergisiso commented on src/psyclone/nemo.py:
```diff
@@ -44,44 +44,89 @@
+    Specialisation of Fparser2ASTProcessor for the Nemo API.
     '''
+    def __init__(self):
+        super(NemoFparser2ASTProcessor, self).__init__()
```
Removed

sergisiso commented on src/psyclone/nemo.py:
```diff
@@ -44,44 +44,89 @@
+        Specialized method to create a NemoLoop instead of a
+        generic Loop.
+
+        :param parent: The parent of the node.
```
OK, I will try to lowercase all `:param:` I get across, but I am sure I still miss some.

sergisiso commented on src/psyclone/nemo.py:
```diff
@@ -44,44 +44,89 @@
 
         :param child: node in fparser2 AST.
         :type child:  :py:class:`fparser.two.utils.Base`
         :param parent: Parent node in the PSyclone IR we are constructing.
```
Done

sergisiso commented on src/psyclone/nemo.py:
```diff
@@ -366,111 +411,18 @@ def ast(self):
-    :type ast: :py:class:`fparser.two.Block_Nonlabel_Do_Construct`
     :param parent: parent of this NemoLoop in the PSyclone AST.
     :type parent: :py:class:`psyclone.psyGen.Node`
     '''
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+
+        if not isinstance(self.children[3], Schedule):
+            raise InternalError(
+                "Loop malformed or incomplete. Forth children should be a "
```
Fixed

arporter replied:
s/children/child/ :-)

sergisiso replied:
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+
+        '''
+        self._check_completeness()
+        self._children[0] = expr
```
Good idea

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+    def stop_expr(self):
+        ''' Return the PSyIR Node representing the Loop stop expression.
+
+        :return: Loop stop expression.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+
+        '''
+        self._check_completeness()
+        self._children[1] = expr
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+    def step_expr(self):
+        ''' Return the PSyIR Node representing the Loop step expression.
+
+        :return: Loop step expression.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+        :type expr: :py:class:`psyclone.psyGen.Node`
+
+        '''
+        self._check_completeness()
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,15 +3030,102 @@ def __init__(self, parent=None,
+    def loop_body(self):
+        ''' Return PSyIR Schedule with the loop body statements.
+
+        :return: PSyIR Schedule with the loop body statements
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -764,13 +771,10 @@ def _lower_bound(self):
-                  "=" + self._id + " lower=" + self._lower_bound() +
-                  "," + self._upper_bound() + "," + step + "\n")
+        result = "Loop[id='" + self._id + "', "
+        result += "variable='" + self._variable_name + "']\n"
```
The idea is that they are already shown in the children below :
```
Loop [...]
    Literal['1']  <- Lower_bound
    Literal['N']  <- Upper bound
    ...
```
However, in the code you reviewed this could have been 'Uninitialised'. I have now make sure that the `self._lower/upper_bound` are called at the beginning of the `__str__`, even if we call them redundantly (but this should be fixed by #440)

Also I am calling now the parent `__str__`

sergisiso commented:
@arporter I addressed the issues raised, it is ready for the next review

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3111,15 +3120,20 @@ def step_expr(self, expr):
-        :return: PSyIR Schedule with the loop body statements
+        '''
+        :return: the PSyIR Schedule with the loop body statements.
         :rtype: list of :py:class:`psyclone.psyGen.Schedule`
```
Not an error from this PR but presumably we don't really return a `list` of `Schedule`s do we?

sergisiso replied:
Good catch

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3309,24 +3322,39 @@ def gen_code(self, parent):
+
+            :return: True if it is equal to the literal '1', false otherwise.
+            '''
+            return isinstance(expr, Literal) and expr.value == '1'
```
Blank line after function definition please.

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3309,24 +3322,39 @@ def gen_code(self, parent):
+            if is_unit_literal(self.step_expr):
+                step_str = None
+            else:
+                import pdb; peb.set_trace()
```
You still have an import of pdb here which must also mean that this else clause is not covered by the test suite.

sergisiso replied:
Ups, I pushed before committing this bit, this is fixed now. Also added a test for the else condition.

arporter commented on src/psyclone/tests/transformations_test.py:
```diff
@@ -126,3 +126,74 @@ def test_ifblock_children_region():
+def test_fusetrans_error_not_same_parent():
+    ''' Check that we reject attempts to fuse loops which don't share the
+    same parent '''
+    from psyclone.psyGen import Loop, Schedule, Literal, Return
```
pylint reports that `Return` is unused.

sergisiso replied:
deleted

arporter commented on src/psyclone/tests/transformations_test.py:
```diff
@@ -126,3 +126,74 @@ def test_ifblock_children_region():
+
+    fuse = LoopFuseTrans()
+
+    # Check first loop
```
Incorrect comment?

sergisiso replied:
fixed

arporter commented on src/psyclone/tests/psyir/backend/fortran_test.py:
```diff
@@ -637,3 +637,26 @@ def test_fw_codeblock():
+        "module test\n"
+        "contains\n"
+        "subroutine tmp()\n"
+        "  integer :: i, sum\n"
```
Strictly speaking we should ensure `sum` is initialised.

sergisiso replied:
done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -838,16 +838,23 @@ def test_ompdo_directive_class_view(capsys):
+                "        " + literal + "[value:'1']\n"
+                "        " + sched + "[]\n"
+                "            " + kern + " testkern_code(a,f1,f2,m1,m2) "
+                "[module_inline=False]\n")
```
Not your mistake but please could you remove the two following `print`s

sergisiso replied:
removed prints

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2696,6 +2690,102 @@ def test_codeblock_can_be_printed():
+        _ = loop.start_expr
+    assert error_str in str(err.value)
+
+    # Expressions that are not PSyIR as not accepted
```
s/as not/are not/

sergisiso replied:
fixed

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1999,11 +2006,11 @@ def test_node_ancestor():
     from psyclone.psyGen import Loop
```
pylint reports that `Loop` is unused.

sergisiso replied:
deleted

arporter commented on src/psyclone/tests/profile_test.py:
```diff
@@ -68,6 +68,7 @@ def teardown_function():
 def test_profile_basic(capsys):
     '''Check basic functionality: node names, schedule view.
     '''
```
pylint reports that GOKern and GOInvokeSchedule are imported into this file but are not used.

sergisiso replied:
deleted imports

arporter commented on src/psyclone/tests/nemo/nemo_psy_test.py:
```diff
@@ -52,14 +52,6 @@
 BASE_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
```
pylint complains that Fortran2003 is imported (line 47) but unused.

sergisiso replied:
deleted

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -168,6 +168,7 @@ def test_loop_fuse_different_iterates_over():
                              newsched.children[1])
 
 
+@pytest.mark.xfail(reason="Needs a new way to produce Unexpected Exception")
```
Please create a new issue or (better) fix this xfail.

sergisiso replied:
Fixed. I didn't realize last time that I had to update to `.children` to `.loop_body.children`

arporter replied:
Great.

arporter commented on src/psyclone/tests/extract_test.py:
```diff
@@ -404,26 +404,16 @@ def test_extract_node_representation(capsys):
-End Schedule""")
-    assert correct in str(schedule)
+    print(str(schedule))
+    assert "\nExtractStart\n" in str(schedule)
```
Maybe also include the text that should be immediately before and after the `ExtractStart`/`ExtractEnd` lines.

sergisiso replied:
As suggested I added the text before and after, but note that the positioning is also checked with the asserts below.

arporter commented on src/psyclone/tests/extract_test.py:
```diff
@@ -404,26 +404,16 @@ def test_extract_node_representation(capsys):
-EndLoop
-End Schedule""")
-    assert correct in str(schedule)
+    print(str(schedule))
```
Please remove this `print`

sergisiso replied:
deleted

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -1024,6 +1024,7 @@ def test_loop_fuse_different_spaces(monkeypatch, dist_mem):
             assert same_space_warning not in str(excinfo.value)
 
 
+@pytest.mark.xfail(reason="Needs a new way to produce Unexpected Exception")
```
Please fix this xfail.

sergisiso replied:
Fixed

sergisiso commented:
@arporter This is ready for next review.

I have addressed the inline comments  and removed the xfails I introduced (also I was happy to see I could remove some xfails from @hiker dependency analysis as he expected, however there is still gocean's that uses derived types and I added a new one with ImplicitLoops, both refer to #440 now)

Unfortunately I see that codecov is again not working properly, I checked the coverage manually but let me know if something is still missing. (in the meantime I will try to manually submit the reports to codecov)

I improved the docstrings that are imported into the Developer's Guide that specify what each children means, is this better? would you prefer an explanation directly there instead of the imported class description?

arporter commented on src/psyclone/nemo.py:
```diff
@@ -569,3 +552,16 @@ def match(node):
         return True
+
+    def reference_accesses(self, var_accesses):
+        '''Get all variable access information. The loop variable is
```
Incorrect docstring - this method doesn't do anything other than call `var_accesses.next_location`. Should the TODO refer to #440?

sergisiso replied:
I updated the docstring and changed the TODO to #440

arporter commented on src/psyclone/tests/dependency_test.py:
```diff
@@ -234,15 +224,36 @@ def test_do_loop_not_working_yet(parser):
 
-@pytest.mark.xfail(reason="Gocean loops boundaries are also strings #400/#444")
+
+@pytest.mark.xfail(reason="Gocean loops boundaries are also strings #440")
```
(Picky) but we should now drop the "also" in this reason.

sergisiso replied:
removed 'also'

arporter commented on doc/developer_guide/developers.rst:
```diff
@@ -363,12 +363,21 @@ However, the necessary nodes in the new tree structure will be annotated
+Iteration construct
+-------------------
+
 .. autoclass:: psyclone.psyGen.Loop
```
This is nice but does reveal some errors in our Sphinx docstrings. In particular, the description of `unique_modified_args` is not formatted correctly and the entry for `reference_accesses` says that the loop variable is marked as READWRITE when I think (although I might be wrong) that it is marked as READ+WRITE. @hiker can probably advise here.

sergisiso replied:
fixed these 2 issues with the docstrings

hiker replied:
Yes, READ+WRITE is now correct, I didn't fix up the documentation, so thanks. Though READ+WRITE is only used in str() to indicate an read and write access at the same location. So maybe this should say that the loop variable is marked with a READ and a WRITE access at the same location?

arporter replied:
Thanks @hiker and @sergisiso.

arporter commented on src/psyclone/nemo.py:
```diff
@@ -49,9 +49,8 @@
     Loop, CodedKern, InternalError, NameSpaceFactory, Schedule, \
-    Fparser2ASTProcessor
+    Fparser2ASTProcessor, colored, SCHEDULE_COLOUR_MAP
 from fparser.two.utils import walk_ast, get_child
```
Not your fault but pylint points out that we already import `walk_ast` and `get_child` and that `copy` and `AccessType` are not used.

sergisiso replied:
Removed unnecessary imports

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3029,9 +3029,16 @@ def coloured_text(self):
-    represents the equivalent to Fortran do loops. This means the loop is
-    bounded by start/stop/step expressions evaluated before the loop starts.)
+    '''
+    Node representing a loop within the PSyIR. It has 4 mandatory children:
```
Fixed

sergisiso commented:
@arporter I fixed the typos, pylint issues and improved suggested docstrings. Ready for next review.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3345,12 +3346,13 @@ def has_inc_arg(self):
                              operator) to search for.
-        :returns: all unique arguments of type arg_type from Kernels in this
-        loop that are modified.
+        :returns: All unique arguments of the given type from kernels inside \
```
Lowercase "all" please.

arporter replied:
Also, lowercase "list" below and no full stop at the end.

sergisiso replied:
fixed

sergisiso commented:
@arporter Fixed docstring formatting and brought to master. This is ready for next review.

arporter commented:
Changes look OK. All tests pass (including with compilation). All (recently) modified lines are covered.
